src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,176 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private int size;
+    private MyNode<T> head;
+    private MyNode<T> tail;
+    private MyNode<T> prev;
+    private MyNode<T> next;
+
     @Override
     public void add(T value) {
+        if (isEmpty()) {
+            createFirstNode(value);
+        } else {
+            createLastNode(value);
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == size || (index == 0 && size == 0)) {
+            add(value);
+            return;
+        }
+
+        if (index == 0 && head != null) {
+            MyNode<T> newNode = new MyNode<>(prev, value, head);
+            head.prev = newNode;
+            head = newNode;
+            size++;
+            return;
+        }
+        checkIndex(index);
+        MyNode<T> searchNode = searchNodeByIndex(index);
+        MyNode<T> oldNode = searchNode.prev;
+        MyNode<T> newNode = new MyNode<>(oldNode, value, searchNode);
+        searchNode.prev = newNode;
+        if (oldNode == null) {
+            head = newNode;
+        } else {
+            oldNode.next = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return searchNodeByIndex(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        MyNode<T> searchNode = searchNodeByIndex(index);
+        T oldElement = searchNode.item;
+        searchNode.item = value;
+        return oldElement;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        return unlink(searchNodeByIndex(index));
     }
 
     @Override
     public boolean remove(T object) {
+        MyNode<T> removeNode = head;
+        if (removeNode.item == object || (object != null && object.equals(removeNode.item))) {
+            remove(0);
+            return true;
+        }
+        for (int i = 1; i < size; i++) {
+            removeNode = removeNode.next;
```

Comment:
let's move this line at the end of the for loop, and then we can remove `if` block at the beginning of the method